python/social_bot/envs/embodied_teacher.py

```diff
@@ -0,0 +1,265 @@
+            resized_image_size=resized_image_size,
+            vocab_sequence_length=vocab_sequence_length)
+            
+        if not self._is_training_for_teacher_policy:
+            # insert teacher model
+            self.insert_model(
+                model=agent_type, name="teacher", pose="0 -2 0 0 0 0")
+            # set up teacher joints
+            self._teacher_joints = []
+            for joint in self._agent_joints:
+                self._teacher_joints.append("teacher::" + joint)
+            self._teacher_embodied = self._world.get_agent('teacher')
+
+    def reset(self):
+        if self._demo_by_human:
+            self._keybo.reset()
+        obs = super().reset()
+        if self._demo_by_human or self._is_training_for_teacher_policy:
+            return obs
+        else:
+            return obs, obs
+
+    def _step_with_teacher_policy(self, teacher_action, agent_action):
+        obs, reward, done, _ = self._step_with_teacher_action(
+            teacher_action, agent_action)
+        teacher_obs = self._get_teacher_obs()
+        return teacher_obs, obs, reward, done, {}
+
+    def _step_with_human_demo(self, agent_action):
+        teacher_action = self._keybo.get_agent_actions(self._agent_type)
+        return self._step_with_teacher_action(teacher_action, agent_action)
+
+    def _step_with_teacher_action(self, teacher_action, agent_action):
+        if self._with_language:
+            sentence = agent_action.get('sentence', None)
+            if type(sentence) != str:
+                sentence = self._teacher.sequence_to_sentence(sentence)
+            controls = agent_action['control']
+        else:
+            sentence = ''
+            controls = agent_action
+        self._take_action(self._agent, self._agent_joints, controls)
+        self._take_action(self._teacher_embodied, self._teacher_joints,
+                          teacher_action)
+        self._world.step(self._sub_steps)
+        teacher_action = self._teacher.teach(sentence)
+        obs = self._get_observation_with_sentence(teacher_action.sentence)
+        self._steps_in_this_episode += 1
+        ctrl_cost = np.sum(np.square(controls)) / controls.shape[0]
+        reward = teacher_action.reward - self._action_cost * ctrl_cost
+        self._cum_reward += reward
+        if teacher_action.done:
+            logging.debug("episode ends at cum reward:" +
+                          str(self._cum_reward))
+        return obs, reward, teacher_action.done, {}
+
+    def _take_action(self, agent, joints, action):
+        controls = np.clip(action, -1.0, 1.0) * self._agent_control_range
+        controls_dict = dict(zip(joints, controls))
+        agent.take_action(controls_dict)
+
+    def _get_teacher_obs(self):
+        agent = self._agent
+        self._agent = self._teacher_embodied
+        for group in self._teacher._task_groups:
+            for task in group._tasks:
+                task._agent = self._teacher_embodied
```

Comment:
This is very hacky. Better to change the function to pass in _teacher_embodied instead.

Reply:
Done